### DISCUSSION TITLE
Drag and drop with OSR

### DISCUSSION THREAD

**Discussion by DanielYampolschi** [CefView/QCefView#361] [General]
Hello, I am trying to implement a drag and drop functionallity on a QCefView client using OSR. (It deoes work out of the box without OSR)
But I have no idea where to start, or what am I supposed to do to make it work.
Could someone please show me, or point,  the way to add this functionallity?
I am getting a bit desperate now, since I tried a lot of things already and I am not making any progress.
Thank you very much.

**Comment by tishion** [comment]
Hi @DanielYampolschi for OSR D&D implementation, you need to consider the two scenarios below:

1. drag data from outside of QCefView and drop it onto QCefView
in this scenario, you don't need to care about where the data comes from, you need to implement Qt [D&D](https://doc.qt.io/qt-6/dnd.html) for QCefView, AKA, at least the following 4 methods:
 - QCefViwe::[QDragEnterEvent](https://doc.qt.io/qt-6/qdragenterevent.html)
 in your implementation, you need to call CefBrowserHost::DragTargetDragEnter
 
 - QCefViwe::[QDragLeaveEvent](https://doc.qt.io/qt-6/qdragleaveevent.html)
 in your implementation, you need to call CefBrowserHost::DragTargetDragLeave
 
 - QCefView::[QDragMoveEvent](https://doc.qt.io/qt-6/qdragmoveevent.html)
  in your implementation, you need to call CefBrowserHost::DragTargetDragOver

- QCefView::[QDropEvent](https://doc.qt.io/qt-6/qdropevent.html)
in your implementation, you need to call CefBrowserHost::DragTargetDrop

obviously, you need to convert the [QMimeData](https://doc.qt.io/qt-6/qmimedata.html) to CefDragData

2. drag data from QCefView and drop it to the outside of QCefView
when the user drag data from QCefView the method CefRenderHandler::StartDragging will get called, if you allow this dragging, you need to create a [QDrag](https://doc.qt.io/qt-6/qdrag.html) operation, please refer to this simple example: https://doc.qt.io/qt-6/dnd.html#dragging


good luck

**Reply by wangtao0505** [reply]
If I want to drag and drop a small image on a web page, how do QMimeData get the image information？
